groups_admin.rb

```diff
@@ -100,34 +125,180 @@ def update_accept_header(extension,mime_type)
 # groups/<group id> (PUT GET DELETE)
 
 ################ groups ###############
-
+# create group representations
+# input:  https://developers.google.com/admin-sdk/directory/v1/guides/manage-groups
+# output: https://developers.google.com/admin-sdk/directory/v1/reference/groups#resource
 
 # list groups
 ## Return list of ids or list of array/hash [id,url]
 ## put / post must accept (optional) settings
-get '/groups' do
-  "get all groups via group email"
+
+get '/groups/?', :provides => [:json, :html] do
+  respond_to do |format|
+    config = {
+        :args => nil,
+        :method_symbol => :list_groups,
+        :handle_result => Proc.new { |result| result },
+        :service_name => 'ADMIN_DIRECTORY'
+    }
+
+    # get the data
+    @data = run_request(config, [])
+    # format as required
+    format.json { erb :'groups.json' }
+    format.html { erb :'groups.json' }
+  end
 end
 
-# create new group / update existing group
-## must accept settings values
+# Expects the body of the put to contain group email, group name and description.
+# If the gid and the email don't agree then
 put '/groups/:gid' do |gid|
-  "create a group: #{gid}"
+  # assumes that request has not already been read
+  body = request.body.read
+  #verify that the email is not in the body (and add) or is the same.
+
+  config = {
+      :args => nil,
+      :method_symbol => :insert_new_group,
+      :handle_result => Proc.new { |result| result },
+      :service_name => 'ADMIN_DIRECTORY'
+  }
+
+  params2 = {:email => params['email'],
+             :name => params['name'],
+             :description => params['description']}
+
+  unless gid == params[:email]
+    halt 422, "requested id is inconsistent"
+  end
+
+  # put argument in array so it isn't treated as a hash of keyword parameters.
+  #run_request(config, [params2])
+
+  begin
+    run_request(config, [params2])
+  rescue GGBServiceAccountError => ggb_err
+    halt ggb_err.status_code, ggb_err.cause.to_json
+  end
+
+end
+
+post '/groups/:gid' do |gid|
```

Comment:
Or "405 Method Not Allowed".  There are additional requirements for that status code, though.  See: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.6